packages/core/src/services/chains/run.ts

```diff
@@ -96,106 +95,120 @@ async function iterate({
   workspace: Workspace
   source: LogSources
   chain: Chain
-  apikeys: CachedApiKeys
+  providersMap: CachedApiKeys
   controller: ReadableStreamDefaultController
   previousCount?: number
-  previousApiKey?: ProviderApiKey
   documentLogUuid: string
-  previousResponse?: ChainTextResponse
+  previousResponse?: ChainStepResponse<StreamType>
   configOverrides?: ConfigOverrides
 }) {
-  try {
-    const step = await computeStepData({
-      chain,
-      previousResponse,
-      apikeys,
-      apiKey: previousApiKey,
-      sentCount: previousCount,
-    })
+  const prevText = previousResponse?.text
+  const stepValidator = new ChainValidator({ prevText, chain, providersMap })
+  const step = await stepValidator.call().then((r) => r.unwrap())
 
-    publishStepStartEvent(controller, step, documentLogUuid)
+  try {
+    const newMessagesInStep = step.conversation.messages.slice(previousCount)
+    const sentCount = previousCount + newMessagesInStep.length
+    publishStepStartEvent(controller, step, newMessagesInStep, documentLogUuid)
 
     const providerProcessor = new ProviderProcessor({
       source,
       documentLogUuid,
       config: step.config,
-      apiProvider: step.apiKey,
+      apiProvider: step.provider,
       messages: step.conversation.messages,
     })
     const aiResult = await ai({
       workspace,
       messages: step.conversation.messages,
       config: step.config,
-      provider: step.apiKey,
+      provider: step.provider,
       schema: getSchemaForAI(step, configOverrides),
       output: getOutputForAI(step, configOverrides),
     })
 
     await streamAIResult(controller, aiResult)
     const response = await providerProcessor.call({
       aiResult,
-      saveSyncProviderLogs: step.completed,
+      saveSyncProviderLogs: step.chainCompleted,
     })
 
-    if (step.completed) {
-      await handleCompletedChain(
-        controller,
-        step,
-        response as ChainCallResponse,
-      )
-      return response as ChainCallResponse
+    if (step.chainCompleted) {
+      await handleCompletedChain(controller, step, response)
+
+      return response
     } else {
-      publishStepCompleteEvent(controller, response as ChainStepResponse)
+      publishStepCompleteEvent(controller, response)
 
       return iterate({
         workspace,
         source,
         chain,
         documentLogUuid,
-        apikeys,
+        providersMap,
         controller,
-        previousApiKey: step.apiKey,
-        previousCount: step.sentCount,
-        previousResponse: response as ChainTextResponse,
+        previousCount: sentCount,
+        previousResponse: response,
         configOverrides,
       })
     }
-  } catch (error) {
-    handleIterationError(controller, error)
+  } catch (e: unknown) {
```

Comment:
can we keep the code in the catch in its own method? this method is hard enough to read already

Reply:
I'm working on that on the next PR